camel/data_collector/sharegpt_collector.py

```diff
@@ -0,0 +1,173 @@
+# ========= Copyright 2023-2024 @ CAMEL-AI.org. All Rights Reserved. =========
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ========= Copyright 2023-2024 @ CAMEL-AI.org. All Rights Reserved. =========
+
+import json
+from typing import Any, ClassVar, Dict, List, Literal, Optional, Union
+
+from pydantic import BaseModel
+from typing_extensions import Self
+
+from camel.agents.chat_agent import ChatAgent
+from camel.data_collector.base import BaseDataCollector
+from camel.messages.base import BaseMessage
+from camel.messages.func_message import FunctionCallingMessage
+from camel.schemas.openai_converter import OpenAISchemaConverter
+from camel.toolkits.function_tool import FunctionTool
+from camel.types.enums import OpenAIBackendRole
+
+# ruff: noqa: E501
+DEFAULT_CONVERTER_PROMPTS = """
+    Extract key entities and attributes from the conversations
+    and convert them into a structured JSON format.
+    For example:
+    System: You are a helpful assistant
+    Tools: [{"name": "get_release_date", "arguments": ["Portal"]}]
+    User: When is the release date of the video game Portal?
+    Assistant: The release date of the video game Portal is October 9, 2007.
+    Your output should be:
+    {
+        "system": "You are a helpful assistant",
+        "tools": "[{"name": "get_release_date", "arguments": ["Portal"]}]",
+        "conversations": [
+            {"from": "human", "value": "When is the release date of the video game Portal?"},
+            {"from": "gpt", "value": "The release date of the video game Portal is October 9, 2007."}
+        ]
+    }
+"""
+
+
+class ConversationItem(BaseModel):
+    from_: Literal["human", "gpt", "function_call", "observation"]
+    value: str
+
+    class Config:
+        fields: ClassVar[Dict[str, str]] = {"from_": "from"}
+        extra = "forbid"
+
+
+class ShareGPTData(BaseModel):
+    system: str
+    tools: str
+    conversations: List[ConversationItem]
+
+    class Config:
+        extra = "forbid"
```

Comment:
could we use the BaseModel defined in camel/messages/conversion/models.py?

Reply:
Why are the system string and tools here? They might vary between conversations. It can be a nice convenience to have a function to get these in the existing ShareGPTConversation (might require some modification to support different role names) if you want to add that (though the system message should still be in the list of messages).

Reply:
Hi Caelum, 

Thank you for your thoughtful feedback! In this design, I referred to the format used in LLaMA-Factory (https://github.com/hiyouga/LLaMA-Factory/blob/main/data/README.md), where tools and the system message are placed in the same position. I was thinking that keeping it consistent with their approach might be a good idea. 

What do you think? I’d be happy to hear your thoughts! 😊

Reply:
Hey @liuxukun2000 , could we unify the model here with `ShareGPTMessage` under `camel/messages/conversion/conversation_models.py`? it's ok to add `system` and `tools` under this but we need to make it optional